pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,278 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/apis/rbac"
+)
+
+const (
+	// APIGroupAll matches all api groups.
+	APIGroupAll = "*"
+	// ResourceAll matches all api resources.
+	ResourceAll = "*"
+	// NonResourceAll matches all api non-resource paths.
+	NonResourceAll = "*"
+	// VerbAll matches all request verbs.
+	VerbAll = "*"
+
+	// GroupKind represents "group" subject kind
+	GroupKind = "Group"
+	// ServiceAccountKind represents "service-account" subject kind
+	ServiceAccountKind = "ServiceAccount"
+	// UserKind represents "user" subject kind
+	UserKind = "User"
+)
+
+// These are valid condition statuses. "ConditionTrue" means a resource is in the condition;
+// "ConditionFalse" means a resource is not in the condition; "ConditionUnknown" means kubernetes
+// can't decide if a resource is in the condition or not. In the future, we could add other
+// intermediate conditions, e.g. ConditionDegraded.
+const (
+	FlowSchemaConditionTrue    FlowSchemaConditionStatus = "True"
+	FlowSchemaConditionFalse   FlowSchemaConditionStatus = "False"
+	FlowSchemaConditionUnknown FlowSchemaConditionStatus = "Unknown"
+
+	PriorityLevelConfigurationConditionTrue    PriorityLevelConfigurationConditionStatus = "True"
+	PriorityLevelConfigurationConditionFalse   PriorityLevelConfigurationConditionStatus = "False"
+	PriorityLevelConfigurationConditionUnknown PriorityLevelConfigurationConditionStatus = "Unknown"
+)
+
+// These are valid flow-dingtinguisher methods.
+const (
+	FlowDistinguisherMethodNone        = "None"
+	FlowDistinguisherMethodByUserName  = "ByUserName"
```

Comment:
BySubject

Reply:
I think "subject" is the generic term, and username is a particular sort of subject.  Others might be based on the user's groups.

Reply:
This needs to work for both users and service accounts in the first take. If "Subject"  includes groups then I agree it won't work. I do not think we should make admins need to know *how* a caller is performing authn. What would "ByUsername" do if the caller has a service account?

I think selecting a particular group should be a separate option.

Reply:
I guess service accounts do have a user name. (I will blame the jet lag.)

Can we just do "ByUser"?

Reply:
I do not understand the remark "I do not think we should make admins need to know how a caller is performing authn"; I do not think I am proposing something contrary to that.  The authentication framework says that authentication produces a username and a set of groups, regardless of the authentication technique used.

https://kubernetes.io/docs/reference/access-authn-authz/authentication/ says that authentication produces a "username" and a set of "groups"; logically it seems to me that both are attributes of the "user" making the request.  So I think that "ByUser" is unclear, but "username" is clear and we may someday define something involving the user's groups.

Reply:
> I guess service accounts do have a user name

yes, it does. the name is in the form of "system:serviceaccounts:<ns>:<sa-name>"

> I do not think we should make admins need to know how a caller is performing authn

i agree that we shouldn't expose more details about authn, at least we haven't done that so far i guess? it'll be hard for users to tell the differences between "User" and "Username"

Reply:
i updated it to `ByUser` now